pkg/reconciler/ingress/resources/virtual_service.go

```diff
@@ -126,42 +126,56 @@ func MakeVirtualServices(ia v1alpha1.IngressAccessor, gateways map[v1alpha1.Ingr
 	}
 
-	// Add probe routes
-	for _, vs := range vss {
-		if _, err := InsertProbe(vs); err != nil {
-			return nil, errors.Wrapf(err, "failed to insert probe route to %s/%s", vs.Namespace, vs.Name)
-		}
-	}
-
 	return vss, nil
 }
 
-// InsertProbe inserts a rule used to probe the VirtualService
-func InsertProbe(vs *v1alpha3.VirtualService) (string, error) {
-	hash, err := computeVirtualServiceHash(vs)
+// InsertProbe adds a AppendHeader rule so that any request going through a Gateway is tagged with
+// the version of the Ingress currently deployed on the Gateway.
+func InsertProbe(ia v1alpha1.IngressAccessor) error {
+	bytes, err := ComputeIngressHash(ia)
 	if err != nil {
-		return "", errors.Wrapf(err, "failed to compute the hash of %s/%s", vs.Namespace, vs.Name)
+		return fmt.Errorf("failed to compute the hash of the IngressAccessor: %v", err)
 	}
-	// As per RFC1153, a DNS label must be under 63 8-bit octets
-	host := fmt.Sprintf("%x", hash) + ProbeHostSuffix
-	vs.Spec.Hosts = append(vs.Spec.Hosts, host)
-	vs.Spec.HTTP = append(vs.Spec.HTTP, makeProbeRoute(host))
-	return host, nil
+	hash := fmt.Sprintf("%x", bytes)
+
+	for _, rule := range ia.GetSpec().Rules {
+		for i := range rule.HTTP.Paths {
+			if rule.HTTP.Paths[i].AppendHeaders == nil {
+				rule.HTTP.Paths[i].AppendHeaders = make(map[string]string)
+			}
+			rule.HTTP.Paths[i].AppendHeaders[network.HashHeaderName] = hash
+		}
+	}
+
+	return nil
 }
 
-// computeVirtualService computes a hash of the VirtualService Spec and Namespace.
-// It is designed to uniquely identify the effect of a VirtualService. Therefore,
-// Name is irrelevant but Namespace is relevant because Gateway resolution is done within the
-// namespace of the VirtualService if a namespace is not specified in the Gateway definition.
-func computeVirtualServiceHash(vs *v1alpha3.VirtualService) ([16]byte, error) {
-	bytes, err := json.Marshal(vs.Spec)
+// ComputeIngressHash computes a hash of the Ingress Spec, Namespace and Name
+func ComputeIngressHash(ia v1alpha1.IngressAccessor) ([16]byte, error) {
+	bytes, err := json.Marshal(ia.GetSpec())
 	if err != nil {
-		return [16]byte{}, fmt.Errorf("failed to serialize the VirtualServiceSpec: %v", err)
+		return [16]byte{}, fmt.Errorf("failed to serialize IngressAccessor: %v", err)
 	}
-	bytes = append(bytes, []byte(vs.Namespace)...)
+	bytes = append(bytes, []byte(ia.GetNamespace())...)
+	bytes = append(bytes, []byte(ia.GetName())...)
+
+	fmt.Printf("Hash: %x, Namespace: %s, Name: %s, JSON: %s\n", md5.Sum(bytes), ia.GetNamespace(), ia.GetName(), string(bytes))
 	return md5.Sum(bytes), nil
 }
 
+// HostsPerGateway returns the set of hosts that each Gateway handles
+func HostsPerGateway(ia v1alpha1.IngressAccessor, gateways map[v1alpha1.IngressVisibility]sets.String) map[string][]string {
+	output := make(map[string][]string)
+	for _, rule := range ia.GetSpec().Rules {
+		for _, host := range expandedHosts(sets.NewString(rule.Hosts...)).List() {
```

Comment:
Do we really need `List()` calls? Since they are O(nlogn)?

Reply:
You mean as opposed to UnsortedList?
Ordering doesn't matter here, so you are right.